### PR TITLE
Add messagesopt option

### PR DESCRIPTION
**Pull request by Shougo**
I have added `messagesopt` option.
Because more prompt can be disabled by `more` option.
Shouldn't the press-enter prompt also be optionally disabled?

### CODE REVIEW

File: src/message.c [hunk 1]
```diff
@@ -1223,118 +1223,124 @@ wait_return(int redraw)
...
+	    {
+		// Put the character back in the typeahead buffer.  Don't use the
+		// stuff buffer, because lmaps wouldn't work.
+		ins_char_typebuf(vgetc_char, vgetc_mod_mask);
+		do_redraw = TRUE;	    // need a redraw even though there is
+					// typeahead
+	    }
+	} else {
+	    c = CAR;
+	    do_sleep(200L, TRUE);
```
Comment (rbtnn) [inline src/message.c]: Why does it sleep 200ms? Please write comments of the reason at this and docs.

Comment (Shougo) [reply src/message.c]: I have added the documentation.

**Comment by Shougo** [comment]
I have renamed option name and it is possible to set wait time.

**Comment by chrisbra** [comment]
Can you please explain the use case here?

**Comment by Shougo** [comment]
The use case is to prevent user key wait to view the messages.
It is like `nomore` option behavior.

Some plugins implements similar feature when output messages.

**Comment by Shougo** [comment]
Hmmm...

```
From test_options.vim:
Found errors in Test_set_all_one_column():
command line..script /home/runner/work/vim/vim/src/testdir/runtest.vim[613]..function RunTheTest[57]..Test_set_all_one_column line 6: Expected 
```

How to fix the error?
I don't get the error when other options.

File: runtime/doc/options.txt [hunk 2]
```diff
@@ -6280,6 +6280,13 @@ A jump table for the options with a short description can be found at |Q_op|.
 	Also see 'copyindent'.
 	Use |:retab| to clean up white space.
 
+					*'pressenterwait'* *'pew'*
+'pressenterwait'	number	(default 0)
+			global
+	When the value is zero or less, the |press-enter| prompt is enabled.
+	When it is less than zero, it means |:sleep| ms is used to allow the
+	user to verify the output
```
Comment (bfrg) [inline runtime/doc/options.txt]: In the second sentence, shouldn't it say: When it is **greater** than zero, it means `|:sleep|` ms is used to allow the user to verify the output.

**Comment by chrisbra** [comment]
Took me a while to understand the option name. Can we think of a better name? Or is everybody fine with the nameing?

**Comment by justinmk** [comment]
The name obviously needs to be improved. Also why not make this a flag on 'shortmess' ?

**Comment by benknoble** [comment]
I still don’t understand the use case (can we not combine echom with sleep, or use popups?), but the English word would be spelled « presenter » (one « s »).

It seems `messagewait` or `messagedelay` might be appropriate, but I can’t figure out why this needs to be in Vim core. Perhaps Shougo can demonstrate some examples of how this can be used and articulate why alternatives are insufficient?

**Comment by clason** [comment]
> the English word would be spelled « presenter » (one « s »).

Yes but that's not the "English word" here; given the context, it's clearly about the "press-enter" prompt -- hence "pressenter".

And the use case is that some people _really_ (and I mean **really**) hate this prompt and will go to the length of patching Vim to burninate it.

(That doesn't mean that this PR is the best way of dealing with this, even if you agree that "press enter" needs to die in a fire. As @justinmk mentioned, it could be folded into existing options; personally, I would prefer to see builtin support for message routing, i.e., distinguish (single-line) messages, long-form output, and explicit confirmation prompts to allow users to deal with them separately.)

**Comment by Shougo** [comment]
> I still don’t understand the use case (can we not combine echom with sleep, or use popups?), but the English word would be spelled « presenter » (one « s »).

It is possible for plugin side.  But it cannot from user side.

> Yes but that's not the "English word" here; given the context, it's clearly about the "press-enter" prompt -- hence "pressenter".

Yes.  It is from Vim wording.

> And the use case is that some people really (and I mean really) hate this prompt and will go to the length of patching Vim to burninate it.

Yes.  I know.  Good reaction is already here.

> Why not make this a flag on 'shortmess' ?

It is possible to add the flag to disable press-enter prompt.  But it cannot define the time.

> Does this truncate messages? Do the non-truncated messages get recorded in :messages history?

I want to implement feature.  But the hit-enter prompt is called after the output is finished, so it cannot save the already output text to the history.

> I would prefer to see builtin support for message routing, i.e., distinguish (single-line) messages, long-form output, and explicit confirmation prompts to allow users to deal with them separately.)

I know.  I think it is the best approach. 

It is already discussed in neovim issues.
https://github.com/neovim/neovim/issues/22478#issuecomment-2480239231
But it is not implemented yet.
I prefer an approach that "works now" over the "best" approach.

File: runtime/doc/options.txt (hunk 2 above)
Comment (Shougo) [reply runtime/doc/options.txt]: Fixed

**Comment by zeertzjq** [comment]
`hit-enter` also appears in Vim documentation

**Comment by chrisbra** [comment]
How about a `'messagesopt'` option. Which can contain the suboptions: `hit-enter/press-enter` and `sleep` And later it can be improved further if we ever decide to add some message routing (as was suggested).

> Why not make this a flag on 'shortmess' ?

Short flag options are a bit cryptic and we are no longer trying to add such options.

**Comment by benknoble** [comment]
> > the English word would be spelled « presenter » (one « s »).
> 
> 
> 
> Yes but that's not the "English word" here; given the context, it's clearly about the "press-enter" prompt -- hence "pressenter".

🤦doh. Thanks for clearing that up! (A better PR description or commit message might have made that more obvious, and perhaps I didn't read closely enough.)

**Comment by Shougo** [comment]
> 🤦doh. Thanks for clearing that up! (A better PR description or commit message might have made that more obvious, and perhaps I didn't read closely enough.)

Sorry.  I thought you all already understood Vim terminology.

> How about a 'messagesopt' option. Which can contain the suboptions: hit-enter/press-enter and sleep And later it can be improved further if we ever decide to add some message routing (as was suggested).

Thank you.  The suboptions means like `viminfo` like option?
I will check the implementation.

In the sub option, what flags are needed?

* wait time?

* enable/disable pressenter?

I would like to add tests to this patch, but I couldn't figure out how to test the behavior of the message area.
I checked for tests of similar behavior, such as 'more' option, but couldn't find any. Could you give me some advice?

**Comment by Shougo** [comment]
Thanks.  What is the behavior if the `wait-return` suboption is not specified and there is no `sleep:` suboption?

**Comment by chrisbra** [comment]
> What is the behavior if the wait-return suboption is not specified and there is no sleep: suboption?

Probably an error.

**Comment by Shougo** [comment]
I have renamed the option name.

I need to fix the tests later...

**Comment by Shougo** [comment]
Hm...  The behavior seems changed.

It is hard to integrate to `messagesopt`.

**Comment by chrisbra** [comment]
What do you mean please?

**Comment by Shougo** [comment]
`msg_hit_enter` flag is not set properly.
If I removed the flag check, it works.

Can you check `messagesopt` and `did_set_messagesopt` implementation?

**Comment by h-east** [comment]
Let's implement it with a little more reference from `'diffopt'`.

1. `did_set_diffopt()` is implemented in optionsstr.c because it is a string option.
Therefore, shouldn't `did_set_messagesopt()` also be implemented in optionsstr.c, not option.c ?
(Perhaps adding it after `did_set_matchpairs()` would be better)

1. Implement `messagesopt_changed()` in message.c and call it from `did_set_messagesopt()`.
(The relationship is similar to `did_set_diffopt()` and `diffopt_changed()`)
(Move the current content of `did_set_messagesopt()` to `messagesopt_changed()`)
(You probably won't need `set_msg_hit_enter()`, `set_msg_wait()`, or `set_msg_history()`. Just assign variables directly, like `diffopt_changed()`)

1. Currently there is only one flag, `hit-enter`, but it would be better to prepare a static variable `msg_flags` for future use.
(Like `diff_flags`)

1. Currently, `did_set_messagesopt()` uses `STRLEN()`, but calling this function multiple times with the same string has a bit processing time cost, so it is better to use the immidiate value.
(Like `diffopt_changed()`)

1. (Although it's not required in this PR) You will probably need to add `expand_set_messagesopt()` by referring to `expand_set_diffopt()`.

**Comment by Shougo** [comment]
Thanks.  I will try it.

> Currently, did_set_messagesopt() uses STRLEN(), but calling this function multiple times with the same string has a bit processing time cost, so it is better to use the immidiate value.
(Like diffopt_changed())

I know that calling STRLEN() is not efficient, but I thought it would be optimized since the value is clearly determined.

**Comment by Shougo** [comment]
I have removed to use `STRLEN()`.

**Comment by zeertzjq** [comment]
Change the constant strings to arrays (or macros) and use `STRLEN_LITERAL`?

**Comment by Shougo** [comment]
> did_set_diffopt() is implemented in optionsstr.c because it is a string option.
Therefore, shouldn't did_set_messagesopt() also be implemented in optionsstr.c, not option.c ?
(Perhaps adding it after did_set_matchpairs() would be better)

Done.

> Implement messagesopt_changed() in message.c and call it from did_set_messagesopt().
(The relationship is similar to did_set_diffopt() and diffopt_changed())
(Move the current content of did_set_messagesopt() to messagesopt_changed())
(You probably won't need set_msg_hit_enter(), set_msg_wait(), or set_msg_history(). Just assign variables directly, like diffopt_changed())

Done.

> Currently there is only one flag, hit-enter, but it would be better to prepare a static variable msg_flags for future use.
(Like diff_flags)

Done.

**Comment by Shougo** [comment]
> Change the constant strings to arrays (or macros) and use STRLEN_LITERAL?

Oh, I have not know `STRLEN_LITERAL`.
But I don't know why the strings should be convert to arrays.
Can you explain it for me?

**Comment by zeertzjq** [comment]
`STRLEN_LITERAL()` only works for arrays or literal strings

**Comment by Shougo** [comment]
> (Although it's not required in this PR) You will probably need to add expand_set_messagesopt() by referring to expand_set_diffopt().

Done.

**Comment by Shougo** [comment]
> STRLEN_LITERAL() only works for arrays and literal strings, not pointers

Thank you.  I get it.

File: src/message.c [hunk 3]
```diff
@@ -1060,14 +1070,95 @@ delete_first_msg(void)
...
+    int		messages_wait_new = 0;
+    int		messages_history_new = 0;
+
+    const char HIT_ENTER[] = "hit-enter";
+    const char WAIT[] = "wait:";
+    const char HISTORY[] = "history:";
+
+    const int LEN_HIT_ENTER = STRLEN_LITERAL(HIT_ENTER);
+    const int LEN_WAIT = STRLEN_LITERAL(WAIT);
+    const int LEN_HISTORY = STRLEN_LITERAL(HISTORY);
```
Comment (h-east) [inline src/message.c]: You are confusing the language specifications of C and C++.
In the case of C language, even if you add "const", it doesn't become a constant.
Every time the function is called, an area is assigned on the stack and the value is set.
You can confirm this by adding the "-S" option at compile time to output the assembler code.

Why don't you simply follow `diffopt_changed()`?
If you have reservations about using immediate values, use a "#define" macro.

Comment (Shougo) [reply src/message.c]: Thanks.  I have removed the constants.

**Comment by Shougo** [comment]
Hm...  I have checked the tests error.

If I added `set messagesopt&` in the tests, the errors are gone.
`messagesopt` initialization works, but `did_set_messagesopt` seems not called.

**Comment by h-east** [comment]
> messagesopt initialization works, but did_set_messagesopt seems not called.

Should that function be called?
Isn't that just speculation or wishful thinking?
Analyze the source code a bit more.
And, If you run Vim on gdb with "watch p_meo", you'll see where `'messagesopt'` is initializing.

**Comment by Shougo** [comment]
> Should that function be called?
> Isn't that just speculation or wishful thinking?

It seems that there is no need to call this function directly because calling this function manually did not change the behavior of the tests.

> And, If you run Vim on gdb with "watch p_meo", you'll see where 'messagesopt' is initializing.

```
Hardware watchpoint 1: p_meo

Old value = (char_u *) 0x0
New value = (char_u *) 0x555555bbd0d1 "hit-enter,history:500"
set_option_default (opt_idx=opt_idx@entry=217, opt_flags=opt_flags@entry=0, compatible=<optimized out>) at option.c:789
warning: Source file is more recent than executable.
789			options[opt_idx].flags &= ~P_ALLOCED;
(gdb) bt
#0  set_option_default (opt_idx=opt_idx@entry=217, opt_flags=opt_flags@entry=0, compatible=<optimized out>) at option.c:789
#1  0x0000555555958acc in set_options_default (opt_flags=opt_flags@entry=0) at option.c:860
#2  0x000055555595f4f7 in set_init_1 (clean_arg=0) at option.c:680
#3  0x0000555555ae7ee5 in common_init (paramp=paramp@entry=0x555555fd7f60 <params>) at main.c:1000
#4  0x000055555577c2da in main (argc=1, argv=<optimized out>) at main.c:190
```

When `set messagesopt=hit-enter,history:100` is executed.

```
:set messagesopt=hit-enter,history:100
Thread 1 "vim" hit Hardware watchpoint 1: p_meo

Old value = (char_u *) 0x555555bbd0d1 "hit-enter,history:500"
New value = (char_u *) 0x5555561fedd0 "hit-enter,history:100"
do_set_option_string (errbuflen=80, opt_idx=217, opt_flags=<optimized out>, argp=<synthetic pointer>, nextchar=<optimized out>, op_arg=OP_NONE,
    flags=<optimized out>, cp_val=<optimized out>, varp_arg=<optimized out>, errbuf=0x7fffffffd1a0 "", value_checked=0x7fffffffd188,
    errmsg=<synthetic pointer>) at option.c:2113
2113	    if (newval == NULL)
(gdb) bt
#0  do_set_option_string
    (errbuflen=80, opt_idx=217, opt_flags=<optimized out>, argp=<synthetic pointer>, nextchar=<optimized out>, op_arg=OP_NONE, flags=<optimized out>, cp_val=<optimized out>, varp_arg=<optimized out>, errbuf=0x7fffffffd1a0 "", value_checked=0x7fffffffd188, errmsg=<synthetic pointer>) at option.c:2113
#1  do_set_option_value
    (errbuflen=80, opt_idx=217, opt_flags=<optimized out>, argp=<synthetic pointer>, prefix=<optimized out>, op=OP_NONE, flags=<optimized out>, varp=<optimized out>, key_name=0x7fffffffd19e "", nextchar=<optimized out>, afterchar=<optimized out>, cp_val=<optimized out>, stopopteval=<synthetic pointer>, errbuf=0x7fffffffd1a0 "") at option.c:2412
#2  do_set_option
    (errbuflen=80, opt_flags=<optimized out>, argp=<synthetic pointer>, arg_start=<optimized out>, startarg=<synthetic pointer>, did_show=<synthetic pointer>, stopopteval=<synthetic pointer>, errbuf=0x7fffffffd1a0 "") at option.c:2628
#3  do_set (arg_start=0x55555618c5e4 "messagesopt=hit-enter,history:100", opt_flags=<optimized out>) at option.c:2708
#4  0x00005555558b33ac in do_one_cmd (cmdlinep=0x7fffffffd370, flags=0, cstack=0x7fffffffd5c0, fgetline=0x5555558c0f30 <getexline>, cookie=0x0)
    at ex_docmd.c:2616
#5  do_cmdline (cmdline=cmdline@entry=0x0, fgetline=<optimized out>, cookie=cookie@entry=0x0, flags=0) at ex_docmd.c:1028
#6  0x0000555555942196 in nv_colon (cap=0x7fffffffdc90) at normal.c:3169
#7  0x0000555555947bfd in normal_cmd (oap=oap@entry=0x7fffffffdd50, toplevel=toplevel@entry=1) at normal.c:952
#8  0x0000555555ae881a in main_loop (cmdwin=cmdwin@entry=0, noexmode=noexmode@entry=0) at main.c:1562
#9  0x0000555555ae9acb in vim_main2 () at main.c:895
#10 0x000055555577c9b4 in main (argc=<optimized out>, argv=<optimized out>) at main.c:441
```

**Comment by Shougo** [comment]
OK.  I get the reason.

`msg_hist_max` is not set until `did_set_messagesopt` is set.
And `did_set_messagesopt` is not called when initialized.
So the command line history feature is broken.

`msg_hist_max` default value should be 500?

**Comment by h-east** [comment]
> It seems that there is no need to call this function directly because calling this function manually did not change the behavior of the tests.

What are you talking about?
You were concerned about the following. Is that right?
```
`did_set_messagesopt()` is not called when executing "set messagesopt&".
```

---

(After reading your new comment:)
Well, if that solved the problem, that's fine, but please explain it a bit more.

**Comment by Shougo** [comment]
>What are you talking about?
>You were concerned about the following. Is that right?

Sorry.  It is my mistake.

I mistakenly thought that `msg_flags` was not set correctly. I commented that
calling `did_set_messagesopt()` before checking `msg_flags` had no effect.

The real issue was not `msg_flags` but `msg_hist_max`. The initial value of
`msg_hist_max` was set to 0, which disabled the command line history.

File: src/message.c [hunk 4]
```diff
@@ -1060,14 +1075,90 @@ delete_first_msg(void)
...
+
+	if (*p != ',' && *p != NUL)
+	{
+	    return FAIL;
+	}
+	if (*p == ',')
+	    ++p;
+    }
+
+    // 'wait' must be set if 'hit-enter' is not exists
```
Comment (zeertzjq) [inline src/message.c]:
```suggestion
    // "wait" must be set if "hit-enter" is not present
```

File: src/message.c [hunk 5]
```diff
@@ -1060,14 +1075,90 @@ delete_first_msg(void)
...
+    if (!(messages_flags_new & MESSAGES_HIT_ENTER)
+	&& !(messages_flags_new & MESSAGES_WAIT))
+    {
+        return FAIL;
+    }
+
+    // 'history' must be set
```
Comment (zeertzjq) [inline src/message.c]:
```suggestion
    // "history" must be set
```

File: src/message.c [hunk 6]
```diff
@@ -1060,14 +1075,90 @@ delete_first_msg(void)
...
+    if (!(messages_flags_new & MESSAGES_HISTORY))
+    {
+        return FAIL;
+    }
+
+    // 'history' must be positive
```
Comment (zeertzjq) [inline src/message.c]:
```suggestion
    // "history" must be positive
```

File: src/message.c [hunk 7]
```diff
@@ -1060,14 +1075,90 @@ delete_first_msg(void)
...
+    if (messages_history_new < 0)
+    {
+        return FAIL;
+    }
+    // 'history' must be <= 10000
```
Comment (zeertzjq) [inline src/message.c]:
```suggestion
    // "history" must be <= 10000
```

**Comment by Shougo** [comment]
Hm...  Why `Vim tags` error exists?

**Comment by h-east** [comment]
You added `call term_sendkeys(buf, ":set messagesopt&\n")` to existing tests. Is this necessary?
I don't think it's necessary if the default values ​​are set appropriately.

**Comment by Shougo** [comment]
> I don't think it's necessary if the default values ​​are set appropriately.

Yes.  It is true.

`msg_flags` must be initialized like `msg_hist_max`.
I have fixed it.

File: src/testdir/test_cmdline.vim [hunk 8]
```diff
@@ -1716,6 +1716,7 @@ func Test_verbose_option()
     command DoSomething echo 'hello' |set ts=4 |let v = '123' |echo v
     call feedkeys("\r", 't') " for the hit-enter prompt
     set verbose=20
+    set messagesopt&
```
Comment (zeertzjq) [inline src/testdir/test_cmdline.vim]: Why is this added here? It shouldn't be needed.

File: src/testdir/test_cmdline.vim [hunk 9]
```diff
@@ -4032,30 +4033,30 @@ func Test_cd_bslash_completion_windows()
   let &shellslash = save_shellslash
 endfunc
 
-func Test_msghistory()
-  " After setting 'msghistory' to 2 and outputting a message 4 times with
-  " :echomsg, is the number of output lines of :messages 2?
-  set msghistory=2
+func Test_messagesopt_history()
+  " After setting messagesopt 'history' to 2 and outputting a message 4 times
```
Comment (zeertzjq) [inline src/testdir/test_cmdline.vim]:
```suggestion
  " After setting 'messagesopt' "history" to 2 and outputting a message 4 times
```

File: src/testdir/test_cmdline.vim [hunk 10]
```diff
@@ -4032,30 +4033,30 @@ func Test_cd_bslash_completion_windows()
...
   echomsg 'foo'
   echomsg 'bar'
   echomsg 'baz'
   echomsg 'foobar'
   call assert_equal(['baz', 'foobar'], GetMessages())
 
-  " When the number of messages is 10 and 'msghistory' is changed to 5, is the
-  " number of output lines of :messages 5?
-  set msghistory=10
+  " When the number of messages is 10 and messagesopt 'history' is changed to
```
Comment (zeertzjq) [inline src/testdir/test_cmdline.vim]:
```suggestion
  " When the number of messages is 10 and 'messagesopt' "history" is changed to
```

File: src/testdir/test_messages.vim [hunk 11]
```diff
@@ -547,6 +548,8 @@ func Test_echowindow()
   CheckScreendump
 
   let lines =<< trim END
+      set messagesopt&
+
```
Comment (zeertzjq) [inline src/testdir/test_messages.vim]: This shouldn't be needed either

File: src/testdir/test_options.vim [hunk 12]
```diff
@@ -2204,6 +2202,7 @@ endfunc
 func Test_opt_reset_scroll()
   CheckRunVimInTerminal
   let vimrc =<< trim [CODE]
+    set messagesopt&
```
Comment (zeertzjq) [inline src/testdir/test_options.vim]: ditto

File: src/message.c [hunk 13]
```diff
@@ -51,6 +52,21 @@ struct msg_hist
...
+#define MESSAGES_WAIT		0x002
+#define MESSAGES_HISTORY	0x004
+
+// args in 'messagesopt' option
+#define MESSAGES_OPT_HIT_ENTER "hit-enter"
+#define MESSAGES_OPT_WAIT "wait:"
+#define MESSAGES_OPT_HISTORY "history:"
+
+// The default is "hit-enter,history:500"
+static int msg_flags = MESSAGES_HIT_ENTER;
```
Comment (zeertzjq) [inline src/message.c]:
```suggestion
static int msg_flags = MESSAGES_HIT_ENTER | MESSAGES_HISTORY;
```

File: src/testdir/test_cmdline.vim (hunk 8 above)
Comment (Shougo) [reply src/testdir/test_cmdline.vim]: Sorry.  It is mistake.  I will remove it later.

File: src/testdir/test_cmdline.vim (hunk 9 above)
Comment (Shougo) [reply src/testdir/test_cmdline.vim]: Fixed.

File: src/testdir/test_cmdline.vim (hunk 10 above)
Comment (Shougo) [reply src/testdir/test_cmdline.vim]: Fixed.

**Comment by Shougo** [comment]
Fixed.

I have added the tests.

File: src/optiondefs.h [hunk 14]
```diff
@@ -1695,6 +1695,9 @@ static struct vimoption options[] =
     {"mesg",	    NULL,   P_BOOL|P_VI_DEF,
 			    (char_u *)NULL, PV_NONE, NULL, NULL,
 			    {(char_u *)FALSE, (char_u *)0L} SCTX_INIT},
+    {"messagesopt","mopt",  P_STRING|P_ALLOCED|P_VI_DEF|P_ONECOMMA|P_COLON|P_NODUP,
```
Comment (zeertzjq) [inline src/optiondefs.h:1698]: I don't think `P_COLON` is correct for this option.
```c
#define P_COLON	     0x80000000L // values use colons to create sublists
```

Comment (zeertzjq) [reply src/optiondefs.h:1698]: Actually never mind. It's correct.

**Comment by Shougo** [comment]
Question: How to test `expand_set_messagesopt()`?

**Comment by zeertzjq** [comment]
Take a look at `Test_set_completion_string_values()` in `test_options.vim`.

File: src/message.c [hunk 15]
```diff
@@ -1060,14 +1076,85 @@ delete_first_msg(void)
...
+	    return FAIL;
+	}
+	if (*p == ',')
+	    ++p;
+    }
+
+    // "wait" must be set if "hit-enter" is not present
+    if (!(messages_flags_new & MESSAGES_HIT_ENTER)
+	&& !(messages_flags_new & MESSAGES_WAIT))
+    {
```
Comment (chrisbra) [inline src/message.c]: style, you can remove the braces here and below for the single statements after the condition

**Comment by chrisbra** [comment]
Thanks. Looking good already, except for minor style issues. Please check @zeertzjq comment regarding expanding of option values. Also can you please squash and rebase? It's getting a bit messy with so many commits.

File: src/message.c (hunk 15 above)
Comment (Shougo) [reply src/message.c]: Fixed.

**Comment by Shougo** [comment]
> Also can you please squash and rebase? It's getting a bit messy with so many commits.

Done.

**Comment by Shougo** [comment]
> Please check @zeertzjq comment regarding expanding of option values.

Done.

File: src/message.c [hunk 16]
```diff
@@ -1060,14 +1076,77 @@ delete_first_msg(void)
...
+
+	if (*p != ',' && *p != NUL)
+	    return FAIL;
+	if (*p == ',')
+	    ++p;
+    }
+
+    // "wait" must be set if "hit-enter" is not present
+    if (!(messages_flags_new & MESSAGES_HIT_ENTER)
+	&& !(messages_flags_new & MESSAGES_WAIT))
```
Comment (h-east) [inline src/message.c]: - Conditions can be summed up into one.
- Comment is a bit inconsistent.

```suggestion
    // Either "wait" or "hit-enter" is required
    if (!(messages_flags_new & (MESSAGES_HIT_ENTER | MESSAGES_WAIT)))
```

File: runtime/doc/options.txt [hunk 17]
```diff
@@ -5634,6 +5634,26 @@ A jump table for the options with a short description can be found at |Q_op|.
 	generated from a list of items, e.g., the Buffers menu.  Changing this
 	option has no direct effect, the menu must be refreshed first.
 
+						*'messagesopt'* *'mopt'*
+'messagesopt' 'mopt'		string	(default "hit-enter,history:500")
```
Comment (h-east) [inline runtime/doc/options.txt]: - Adjusted tab.

```suggestion
'messagesopt' 'mopt'	string	(default "hit-enter,history:500")
```

File: runtime/doc/options.txt [hunk 18]
```diff
@@ -5634,6 +5634,26 @@ A jump table for the options with a short description can be found at |Q_op|.
...
+			global
+
+	Option settings when outputting messages.  It can consist of the following
+	items.  Items must be separated by a comma.
```
Comment (h-east) [inline runtime/doc/options.txt]: - Fixed exceeding 80 columns.

```suggestion
	Option settings when outputting messages.  It can consist of the
	following items.  Items must be separated by a comma.
```

File: runtime/doc/options.txt [hunk 19]
```diff
@@ -5634,6 +5634,26 @@ A jump table for the options with a short description can be found at |Q_op|.
...
+			'cmdheight' size.
+
+	wait:{n}	Ignored when "hit-enter" is present. Instead of using
+			|hit-enter| prompt, will simply wait for {n}
+			milliseconds so the user has a chance to read the
+			message, use 0 to disable sleep (but then
+			the user may miss an important message).
+
+	history:{n}	Determines how many entries are remembered in the
+			|:messages| history. The maximum value is 10000.
```
Comment (h-east) [inline runtime/doc/options.txt]: - Tweaked to "two-space convention".

```suggestion
			|:messages| history.  The maximum value is 10000.
```

File: runtime/doc/options.txt [hunk 20]
```diff
@@ -5634,6 +5634,26 @@ A jump table for the options with a short description can be found at |Q_op|.
...
+'messagesopt' 'mopt'		string	(default "hit-enter,history:500")
+			global
+
+	Option settings when outputting messages.  It can consist of the following
+	items.  Items must be separated by a comma.
+
+	hit-enter	Use |hit-enter| prompt when the message is longer than
+			'cmdheight' size.
+
+	wait:{n}	Ignored when "hit-enter" is present. Instead of using
```
Comment (h-east) [inline runtime/doc/options.txt]: - Tweaked to "two-space convention".

```suggestion
	wait:{n}	Ignored when "hit-enter" is present.  Instead of using
```

File: runtime/doc/options.txt [hunk 21]
```diff
@@ -5634,6 +5634,26 @@ A jump table for the options with a short description can be found at |Q_op|.
...
+			|hit-enter| prompt, will simply wait for {n}
+			milliseconds so the user has a chance to read the
+			message, use 0 to disable sleep (but then
+			the user may miss an important message).
```
Comment (h-east) [inline runtime/doc/options.txt]: - Fixed line break position.

```suggestion
			message, use 0 to disable sleep (but then the user may
			miss an important message).
```

**Comment by chrisbra** [comment]
thanks all!

**Comment by Shougo** [comment]
Thank you!

**Comment by h-east** [comment]
BTW, "wait:" does not have a maximum value set.
Shouldn't we set an upper limit?

Below, an overflow occurs, but no error and there is no wait.
```vim
:set messagesopt=history:200,wait:9999999999999999999999999999999999999
:echomsg "AAABBBB" | echomsg "XXXXX" | echomsg "ZZZZZZZZZZZZZZZZZZZ"
```

**Comment by Shougo** [comment]
Oh....

It should max value.

**Comment by justinmk** [comment]
Did anyone try `:set messagesopt-=hit-enter` ? 

```
:set messagesopt-=hit-enter
E474: Invalid argument: messagesopt-=hit-enter
```

**Comment by zeertzjq** [comment]
Maybe it can truncate the message if neither `hit-enter` nor `wait:` is present?

**Comment by h-east** [comment]
The document seems inappropriate.
Need to specify either "hit-enter" or "wait:", or both.
Therefore, if you want to switch from "hit-enter" to "wait:" using `+=` and/or `-=`, you would do it as follows.

```vim
:set messagesopt+=wait:3000
:set messagesopt-=hit-enter
```

**Comment by ychin** [comment]
I have been playing around with this a little bit. It seems like right now it's calling `do_sleep` for the timeout? I find that it's pretty annoying if you set the timeout to anything more than a couple hundred milliseconds. Here's my issue with it:

1. You *have* to press Ctrl-C to cancel out of the pause if you don't want to wait. Let's say you configured the timeout to be 2 seconds (which is still barely enough time to read), now Vim will randomly freeze and you have to just know about the Ctrl-C trick because pressing Enter doesn't clear it. I think it's a pretty unintuitive interface. (Meanwhile your input are cached, so after the timeout ends suddenly all your inputs take effect)
2. In GUI Vim, pressing Ctrl-C to cancel out of a sleep is not immediate, as there is a 1-second delay (in terminal the sleep is stopped by an interrupt sent by the terminal but I don't think we do that for GVim). This leads to an even more sluggish behavior than having to press enter (which while annoying is at least immediate).
3. If the timeout is too short you barely have time to read the message. If the timeout is too long, then you essentially freeze Vim and have to dismiss it, which makes it no better than having to press enter. The message is also not stored anywhere so you are really rolling a dice in terms of your timeout configuration.

For (1) and (2), I think the fix is to wait for input rather than calling `do_sleep()`, similar to the "more" prompt. This will allow us to handle other keys like Enter to dismiss the prompt. We can file a PR to fix that.

For (3) though, i think this is where I'm genuinely confused. Are people actually using this with non-zero wait time? How is this different from just disabling hit-enter completely (which is a somewhat blunt tool to use given the many hit-enter scenarios there are in Vim)? I think the issue to me is that the ergonomics of any timeout higher than 1 sec is just worse than hit enter, IMO. There's also no way to recall the now dismissed message, and some hit-enter messages are definitely higher value than others so it feels bad when a message just disappears into the ether without a chance to look at it first. I do agree with @justinmk that this seems like a rather blunt way to handle the issue, whereas agreeing on a more flexible message routing system seems like the way to go here. I guess some people just want to kill hit-enter, but I'm just worried that this setting wouldn't be as useful if someone actually wants an option to read the output, occasionally.

**Comment by Shougo** [comment]
Hi, I am the author.

> For (1) and (2), I think the fix is to wait for input rather than calling do_sleep(), similar to the "more" prompt. This will allow us to handle other keys like Enter to dismiss the prompt. We can file a PR to fix that.

It seems good.  Can you create the PR?

> For (3) though, i think this is where I'm genuinely confused. Are people actually using this with non-zero wait time? How is this different from just disabling hit-enter completely (which is a somewhat blunt tool to use given the many hit-enter scenarios there are in Vim)? I think the issue to me is that the ergonomics of any timeout higher than 1 sec is just worse than hit enter, IMO. There's also no way to recall the now dismissed message, and some hit-enter messages are definitely higher value than others so it feels bad when a message just disappears into the ether without a chance to look at it first. I do agree with @justinmk that this seems like a rather blunt way to handle the issue, whereas agreeing on a more flexible message routing system seems like the way to go here. I guess some people just want to kill hit-enter, but I'm just worried that this setting wouldn't be as useful if someone actually wants an option to read the output, occasionally.

Regarding the issue of not being able to read the command output, I believe this can be addressed by making such settings.

https://github.com/vim/vim/issues/16581#issuecomment-2635636376

In the first place, truly important messages should remain in `:messages`.
I agree that adding messages API would be a fundamental solution to the problem.

**Comment by ychin** [comment]
> It seems good. Can you create the PR?

Yes I can make that PR. Don't have time to do so immediately, but I'll get to it shortly. Need to think through exactly what the desired behavior is.

There are also some commonality between this and `:echowin` as well. I almost think we should just give people an option to pipe commands directly to `:echowin` but I think the issue is we probably only want to do that if the message ends up being too long to fit and we may not have that information (it's been a while since I looked at that part of the code), unless we refactor the core around to store it.
